limitador/src/errors.rs

```diff
@@ -1,14 +1,51 @@
+use crate::limit::ConditionParsingError;
 use crate::storage::StorageErr;
-use thiserror::Error;
+use std::convert::Infallible;
+use std::error::Error;
+use std::fmt::{Display, Formatter};
 
-#[derive(Error, Debug, Eq, PartialEq)]
+#[derive(Debug)]
 pub enum LimitadorError {
-    #[error("error while accessing the limits storage: {0:?}")]
-    Storage(String),
+    StorageError(StorageErr),
+    InterpreterError(ConditionParsingError),
+}
+
+impl Display for LimitadorError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            LimitadorError::StorageError(err) => {
+                write!(f, "error while accessing the limits storage: {err:?}")
+            }
+            LimitadorError::InterpreterError(err) => {
+                write!(f, "error parsing condition: {err:?}")
+            }
+        }
+    }
+}
+
+impl Error for LimitadorError {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        match self {
+            LimitadorError::StorageError(err) => Some(err),
+            LimitadorError::InterpreterError(err) => Some(err),
+        }
+    }
 }
 
 impl From<StorageErr> for LimitadorError {
     fn from(e: StorageErr) -> Self {
-        Self::Storage(e.msg().to_owned())
+        Self::StorageError(e)
+    }
+}
+
+impl From<ConditionParsingError> for LimitadorError {
+    fn from(err: ConditionParsingError) -> Self {
+        LimitadorError::InterpreterError(err)
+    }
+}
+
+impl From<Infallible> for LimitadorError {
```

Comment:
nit: 
out of curiosity, why do we need to add this implementation?

When I apply this patch, it is not needed.


```diff
diff --git a/limitador/src/errors.rs b/limitador/src/errors.rs
index d590aaf..3de262e 100644
--- a/limitador/src/errors.rs
+++ b/limitador/src/errors.rs
@@ -1,6 +1,5 @@
 use crate::limit::ConditionParsingError;
 use crate::storage::StorageErr;
-use std::convert::Infallible;
 use std::error::Error;
 use std::fmt::{Display, Formatter};
 
@@ -43,9 +42,3 @@ impl From<ConditionParsingError> for LimitadorError {
         LimitadorError::InterpreterError(err)
     }
 }
-
-impl From<Infallible> for LimitadorError {
-    fn from(value: Infallible) -> Self {
-        unreachable!("unexpected infallible value: {:?}", value)
-    }
-}
diff --git a/limitador/src/limit.rs b/limitador/src/limit.rs
index 94b95b1..511f083 100644
--- a/limitador/src/limit.rs
+++ b/limitador/src/limit.rs
@@ -1080,7 +1080,7 @@ mod tests {
             limit1.namespace.clone(),
             limit1.max_value + 10,
             limit1.seconds,
-            limit1.conditions.clone(),
+            vec!["req.method == 'GET'"],
             limit1.variables.clone(),
         )
         .expect("This must be a valid limit!");
```

Not arguing to change it, just curiosity.